### PR TITLE
Cosmic Ascension Balance Patch

### PR DESCRIPTION
**Pull request by Comxy**
## About The Pull Request
Now the stargazer does more damage to people in the area and prioritizes mobs attacking them instead of random circuit bots 7 tiles away. It will also slowly heal you. Makes some abilities create more carpet fields when ascending. Increases combo times when ascending. Also makes the Star Gazer die only when the owner dies. Star Gazer will explode when it does die. Star Gazer has a damage aura that does heal the owner.
## Why It's Good For The Game
I have seen a lot of people complaining that the cosmic ascension isn't good and not worth getting so I thought I would make it a little better.
## Changelog
:cl:
balance: Star Gazer now explodes when it dies.
balance: Star Gazer will only die when the owner dies.
balance: Star Gazer will attack the most recent enemy that attacked him when on autopilot.
balance: Star Gazer has an aura that slowly damages people and heals the owner.
balance: Cosmic Ascension now buffs some spells.
/:cl:

### CODE REVIEW

**Comment by Helg2** [comment]
Won't it be better if star gazer would be sentient?

**Comment by Comxy** [comment]
Maybe, but I want this to work out as best as possible

**Comment by Voudez** [comment]
I guess, the best thing to buff a Star gazer is make him sentient like other herecy summons, since it's mostly not about a damage or focusing not dangerous targets, its about standing still, and being surrounded by 3-8 melee armed people, wich allows kill him under 2 minutes. Its also behaves like pet, right? its a bit hard to give him commands, then point at someone, who it should attack, especially during combat.

**Comment by DaydreamIQ** [comment]
The star gazer is really cool but the issue is just how unintuitive it is to yell "attack!" Then point at your victim of choice, whereas if it was sentient / player controlled you could just rush down the target in half the time

**Comment by Looks-to-the-Moon** [comment]
In testing, I found that the best way to control Star Gazer was by space jaunting and telling it commands. Saying either 'follow' to position it wherever, or 'free' to allow it to go ham. If you dont free it he just sits there. You dont need to point and say attack, it will attack on its own if you tell it to be free.

Also, if you point at him with the attack command, he will melee himself endlessly.

Keep in mind this was only in testing.

If you could alt click him at range to command it might be better.

**Comment by KittyNoodle** [comment]
I honestly don’t think he should be killable, and should instead despawn when the summoner dies.

**Comment by ATH1909** [comment]
it might be nice to maybe give some action buttons/spells that replicate commands you can give to him, since our current pet command system might be unintuitive for an uninformed player to use/understand fully.

**Comment by Jacquerel** [comment]
A button which opens the command wheel from an arbitrary range would probably be useful, although you'd probably also need to override the commands it gets to remove any line of sight checks it performs.

**Comment by nukularpower** [comment]
when will rust ascend ever made not poo.  but very nice on this one

**Comment by Helg2** [comment]
Can you make fire shark ritual little harder to do? 9/10 someone picks cosmic heretic its just spamming of fire sharks.

**Comment by nukularpower** [comment]
I've seen a lot of people complaining about the gazer not being sentient.  I think they are going to be proven vastly wrong - if you've explored space lately you know how vicious mobs have become lately (see the recent upsurge in viscerator usage for more examples.)   Also note the prevalance of on-station carp attacks

This thing is going to be OP as hell on worm level, in my opinion, and Rust ascend is really the one that needs help :/

File: code/datums/elements/ai_target_damagesource.dm [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+		return ELEMENT_INCOMPATIBLE
+
+	target.AddElement(/datum/element/relay_attackers)
+	RegisterSignal(target, COMSIG_ATOM_WAS_ATTACKED, PROC_REF(on_attacked))
+
+/datum/element/ai_target_damagesource/Detach(datum/source, ...)
+	. = ..()
+	UnregisterSignal(source, COMSIG_ATOM_WAS_ATTACKED)
+
+/// Add an attacking atom to a blackboard list of things which attacked us
```
Comment (Jacquerel) [inline code/datums/elements/ai_target_damagesource.dm:19]: Not true any more, update these comments.

File: code/datums/elements/death_explosion.dm [hunk 2]
```diff
@@ -0,0 +1,37 @@
+/**
+ * ## death explosion element!
+ *
+ * Bespoke element that spawns one type of gas when a mob is killed
```
Comment (Jacquerel) [inline code/datums/elements/death_explosion.dm:4]: It doesn't do that, update this comment.

File: code/datums/elements/death_explosion.dm [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	src.devastation = devastation
+	src.heavy_impact = heavy_impact
+	src.light_impact = light_impact
+	RegisterSignal(target, COMSIG_LIVING_DEATH, PROC_REF(on_death))
+
+/datum/element/death_explosion/Detach(datum/target)
+	. = ..()
+	UnregisterSignal(target, COMSIG_LIVING_DEATH)
+
+///signal called by the stat of the target changing
```
Comment (Jacquerel) [inline code/datums/elements/death_explosion.dm:29]:
```suggestion
/// Triggered when target dies, make an explosion.
```

File: code/datums/elements/death_linked.dm [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+/datum/element/death_linked/Detach(datum/target)
+	. = ..()
+	UnregisterSignal(target, COMSIG_LIVING_DEATH)
+
+///signal called by the stat of the target changing
+/datum/element/death_linked/proc/on_death(mob/living/target, gibbed)
+	SIGNAL_HANDLER
+	var/mob/living/linked_mob_unresolved = linked_mob?.resolve()
+	if(linked_mob_unresolved)
+		linked_mob_unresolved.death(TRUE)
```
Comment (Jacquerel) [inline code/datums/elements/death_linked.dm:30]:
```suggestion
	var/mob/living/linked_mob_resolved = linked_mob?.resolve()
	linked_mob_resolved?.death(gibbed = TRUE)
```
If it were unresolved we couldn't do any operations on it

File: code/modules/antagonists/heretic/knowledge/cosmic_lore.dm [hunk 5]
```diff
@@ -243,10 +250,28 @@
...
+	star_gazer_mob.AddComponent(/datum/component/damage_aura, range = 7, burn_damage = 0.5, simple_damage = 0.5, has_owner = user)
 	star_gazer_mob.befriend(user)
+	var/datum/action/cooldown/open_mob_commands/commands_action = new /datum/action/cooldown/open_mob_commands()
+	commands_action.Grant(user, star_gazer_mob)
 	var/datum/action/cooldown/spell/touch/star_touch/star_touch_spell = locate() in user.actions
 	if(star_touch_spell)
 		star_touch_spell.set_star_gazer(star_gazer_mob)
+		star_touch_spell.ascended = TRUE
+
+	var/datum/antagonist/heretic/heretic_datum = IS_HERETIC(user)
```
Comment (Jacquerel) [inline code/modules/antagonists/heretic/knowledge/cosmic_lore.dm:266]: This works but don't do it, it's ugly 😔 
While very unlikely it's always possible that some day someone will refactor this macro to return a boolean like it implies that it does and then your code will break.
```suggestion
	var/datum/antagonist/heretic/heretic_datum = user.mind.has_antag_datum(/datum/antagonist/heretic)
```

**Review by Jacquerel — CHANGES_REQUESTED** [review]
We let this go kind of long without real review huh? Sorry about that.
Can you please update the PR body and changelog to reflect in more detail what changes you actually made? They seem to have grown since the initial posting.

File: code/datums/components/damage_aura.dm [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+	var/list/organ_damage = null
+
+	/// Amount of damage to damage on simple mobs over a second
+	var/simple_damage = 0
+
+	/// Which factions are immune to the damage aura
+	var/list/immune_factions = null
+
+	/// Sets a special set of conditions for the owner
+	var/datum/weakref/current_owner = null
```
Comment (Jacquerel) [inline code/datums/components/damage_aura.dm:43]: I still don't really love the ambiguity between "owner" (of this component) and "owner" (abstractly of an arbitrary atom which this component could be attached to) but I also don't have a good suggestion for an alternate name.

Comment (Comxy) [reply code/datums/components/damage_aura.dm:43]: Yeah, maybe special_target I thought about now, but oh well

**Comment by nukularpower** [comment]
After seeing this in action a few times I will once again stress that the people who thought this was weak were very wrong, and it is in fact quite overpowered - it does not need more buffs, and in fact, probably the opposite

also rusthereticbuffwhen

**Pull request merged**